modules/gateway/peers.go

```diff
@@ -207,18 +292,41 @@ func (g *Gateway) threadedAcceptConn(conn net.Conn) {
 		delete(g.peers, kick)
 		g.log.Printf("INFO: disconnected from %v to make room for %v", kick, addr)
 	}
-	// add the peer
-	g.addPeer(&peer{
-		Peer: modules.Peer{
-			NetAddress: addr,
-			Inbound:    true,
-			Version:    remoteVersion,
-		},
-		sess: muxado.Server(conn),
-	})
-	g.mu.Unlock()
+}
+
+// acceptConnPortHandshake performs the port handshake and should be called on
+// the side accepting a connection request. The remote address is only returned
+// if err == nil.
+func acceptConnPortHandshake(conn net.Conn) (remoteAddr modules.NetAddress, err error) {
+	addr := modules.NetAddress(conn.RemoteAddr().String())
+
+	// Read the peer's port that we can dial them back on.
+	var dialbackPort string
+	err = encoding.ReadObject(conn, &dialbackPort, 13) // Max port # is 65535 (5 digits long) + 8 byte string length prefix
+	if err != nil {
+		return "", fmt.Errorf("could not read remote peer's (%v) port: %v", addr, err)
+	}
+	if _, err := strconv.Atoi(dialbackPort); err != nil {
+		return "", fmt.Errorf("peer (%v) sent an invalid dialback port: %v", addr, err)
+	}
+	remoteAddr = modules.NetAddress(net.JoinHostPort(addr.Host(), dialbackPort))
+	// This check should only fail if dialbackPort == 0, which we could check
+	// manually, but there's no harm in validating the entire address.
+	if err := remoteAddr.IsValid(); err != nil {
```

Comment:
The string is sent as a port and I was worried that an adversary could send a port string that changed the host. Something similar to the peer sending the port "::x:9981" to which would append "::x" to the host portion of `remoteAddr`. This specific example isn't possible because IPV6 addresses must be enclosed with square brackets. It's probably overkill, but that's why the check is there.

Alternatively we replace the `Atoi` check with `if remoteAddr.Host() != addr.Host()`. Or instead of sending a port string we send a port integer.